## Machine Learning Pipeline

As the class practice, the students will be required to develop a machine learning pipeline using the `Churn_Modelling_train_test.csv` dataset.

**About dataset**

This dataset is obained from [kaggle](https://www.kaggle.com/datasets/shubhammeshram579/bank-customer-churn-prediction?resource=download). It contains information on bank customers who either left the bank or continue to be a customer. The dataset includes the following attributes:

* Customer ID: A unique identifier for each customer
* Surname: The customer's surname or last name
* Credit Score: A numerical value representing the customer's credit score
* Geography: The country where the customer resides (France, Spain or Germany)
* Gender: The customer's gender (Male or Female)
* Age: The customer's age.
* Tenure: The number of years the customer has been with the bank
* Balance: The customer's account balance
* NumOfProducts: The number of bank products the customer uses (e.g., savings account, credit card)
* HasCrCard: Whether the customer has a credit card (1 = yes, 0 = no)
* IsActiveMember: Whether the customer is an active member (1 = yes, 0 = no)
* EstimatedSalary: The estimated salary of the customer
* Exited: Whether the customer has churned (1 = yes, 0 = no)


### Exploratory Data Analysis

In this section the students are required to do an EDA to understand the dataset.

In [3]:
# import packages
import os
import pandas as pd 

print(os.getcwd())
print(os.listdir("datasets"))

/Users/aaronglanz/Documents/GitHub/mlops-and-system-design/session_2
['Churn_Modelling_train_test.csv', 'bank-full_train_test.csv']


In [4]:
# Load the dataset
...

df = pd.read_csv("datasets/Churn_Modelling_train_test.csv")
df.head()

,RowNumber,CustomerId,Surname,CreditScore,Geography,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited
0,4784,15729224,Jennings,710,France,Female,37.0,5,0.00,2,1.0,0.0,115403.31,0
1,1497,15799156,Okwuadigbo,569,Spain,Male,38.0,8,0.00,2,0.0,0.0,79618.79,0
2,1958,15674922,Beavers,710,France,Male,54.0,6,171137.62,1,1.0,1.0,167023.95,1
3,9174,15653572,Thornton,673,Spain,Male,43.0,8,127132.96,1,0.0,1.0,6009.27,1
4,9748,15775761,Iweobiegbunam,610,Germany,Female,69.0,5,86038.21,3,0.0,0.0,192743.06,1


In [5]:
# dataset analysis
print(df.shape)
display(df.head())
display(df.info())
display(df.describe())
print(df.isnull().sum())
print(df.duplicated().sum())

(9001, 14)


,RowNumber,CustomerId,Surname,CreditScore,Geography,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited
0,4784,15729224,Jennings,710,France,Female,37.0,5,0.00,2,1.0,0.0,115403.31,0
1,1497,15799156,Okwuadigbo,569,Spain,Male,38.0,8,0.00,2,0.0,0.0,79618.79,0
2,1958,15674922,Beavers,710,France,Male,54.0,6,171137.62,1,1.0,1.0,167023.95,1
3,9174,15653572,Thornton,673,Spain,Male,43.0,8,127132.96,1,0.0,1.0,6009.27,1
4,9748,15775761,Iweobiegbunam,610,Germany,Female,69.0,5,86038.21,3,0.0,0.0,192743.06,1


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9001 entries, 0 to 9000
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        9001 non-null   int64  
 1   CustomerId       9001 non-null   int64  
 2   Surname          9001 non-null   object 
 3   CreditScore      9001 non-null   int64  
 4   Geography        9000 non-null   object 
 5   Gender           9001 non-null   object 
 6   Age              9000 non-null   float64
 7   Tenure           9001 non-null   int64  
 8   Balance          9001 non-null   float64
 9   NumOfProducts    9001 non-null   int64  
 10  HasCrCard        9000 non-null   float64
 11  IsActiveMember   9001 non-null   float64
 12  EstimatedSalary  9001 non-null   float64
 13  Exited           9001 non-null   int64  
dtypes: float64(5), int64(6), object(3)
memory usage: 984.6+ KB


None

,RowNumber,CustomerId,CreditScore,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited
count,9001.000000,9.001000e+03,9001.000000,9000.000000,9001.000000,9001.000000,9001.000000,9000.000000,9001.000000,9001.000000,9001.000000
mean,5003.366515,1.569072e+07,650.681369,38.901781,5.007888,76222.210827,1.531497,0.705667,0.514721,100180.823967,0.204422
std,2884.787499,7.179353e+04,96.539591,10.450760,2.894025,62432.198151,0.579398,0.455768,0.499811,57561.189534,0.403301
min,2.000000,1.556570e+07,350.000000,18.000000,0.000000,0.000000,1.000000,0.000000,0.000000,11.580000,0.000000
25%,2509.000000,1.562811e+07,584.000000,32.000000,2.000000,0.000000,1.000000,0.000000,0.000000,50972.600000,0.000000
50%,4989.000000,1.569096e+07,652.000000,37.000000,5.000000,96997.090000,1.000000,1.000000,1.000000,100556.980000,0.000000
75%,7505.000000,1.575265e+07,718.000000,44.000000,7.000000,127450.140000,2.000000,1.000000,1.000000,149458.730000,0.000000
max,10000.000000,1.581569e+07,850.000000,92.000000,10.000000,250898.090000,4.000000,1.000000,1.000000,199992.480000,1.000000


RowNumber          0
CustomerId         0
Surname            0
CreditScore        0
Geography          1
Gender             0
Age                1
Tenure             0
Balance            0
NumOfProducts      0
HasCrCard          1
IsActiveMember     0
EstimatedSalary    0
Exited             0
dtype: int64
2


In [6]:
# review categorical variables
categorical_cols = ["Geography", "Gender"]

for col in categorical_cols:
    print(f"\n{col}")
    print(df[col].value_counts())
    print(df[col].value_counts(normalize=True))


Geography
Geography
France     4505
Spain      2250
Germany    2245
Name: count, dtype: int64
Geography
France     0.500556
Spain      0.250000
Germany    0.249444
Name: proportion, dtype: float64

Gender
Gender
Male      4906
Female    4095
Name: count, dtype: int64
Gender
Male      0.545051
Female    0.454949
Name: proportion, dtype: float64


In [7]:
# review binary variables
binary_cols = ["HasCrCard", "IsActiveMember", "Exited"]

for col in binary_cols:
    print(f"\n{col}")
    print(df[col].value_counts())
    print(df[col].value_counts(normalize=True))


HasCrCard
HasCrCard
1.0    6351
0.0    2649
Name: count, dtype: int64
HasCrCard
1.0    0.705667
0.0    0.294333
Name: proportion, dtype: float64

IsActiveMember
IsActiveMember
1.0    4633
0.0    4368
Name: count, dtype: int64
IsActiveMember
1.0    0.514721
0.0    0.485279
Name: proportion, dtype: float64

Exited
Exited
0    7161
1    1840
Name: count, dtype: int64
Exited
0    0.795578
1    0.204422
Name: proportion, dtype: float64


In [8]:
# review numerical variables
numerical_cols = [
    "CreditScore", "Age", "Tenure", "Balance",
    "NumOfProducts", "EstimatedSalary"
]

df[numerical_cols].describe()

,CreditScore,Age,Tenure,Balance,NumOfProducts,EstimatedSalary
count,9001.000000,9000.000000,9001.000000,9001.000000,9001.000000,9001.000000
mean,650.681369,38.901781,5.007888,76222.210827,1.531497,100180.823967
std,96.539591,10.450760,2.894025,62432.198151,0.579398,57561.189534
min,350.000000,18.000000,0.000000,0.000000,1.000000,11.580000
25%,584.000000,32.000000,2.000000,0.000000,1.000000,50972.600000
50%,652.000000,37.000000,5.000000,96997.090000,1.000000,100556.980000
75%,718.000000,44.000000,7.000000,127450.140000,2.000000,149458.730000
max,850.000000,92.000000,10.000000,250898.090000,4.000000,199992.480000


**EDA Conclusions**

From exploring the dataset we can see the following information:
The dataset contains bank customer information with a target target variable of `Exited`, which indicates whether a customer left the bank(customers that have churned). There are no  missing values or duplicate rows. The dataset includes categorical variables such as `Geography` and `Gender`, binary variables such as `HasCrCard`, `IsActiveMember`, and `Exited`, and numerical variables such as `CreditScore`, `Age`, `Balance`, `NumOfProducts`, and `EstimatedSalary`.

Columns such as `RowNumber`, `CustomerId`, and `Surname` are identifiers and should not be used for model training because they do not provide meaningful predictive information. The target variable should be checked for imbalance before modelling, because if most customers did not churn, a model could achieve high accuracy by mostly predicting the majority class.

The average(mean) customer age is around 39 years old, with the lowest age being 18 and the highest being 92. The average salary is 101k with the lowest being 11.5k and the hights beeing 200k. Average credit score i 650, lowest is 350 and highest is 850. Average tenure of 5 years and average balance of 76k. 

### ML Pipeline

In this section the students are required to create a ML pipeline to predict whether the customer left the bank or not.

In [ ]:
# import packages
import pandas as pd

In [9]:
# Preprocess the features - create a function or a class for it 
def transform(df: pd.DataFrame) -> pd.DataFrame:
    df = df.copy()
    
    # Drop columns that are IDs or not useful for prediction
    df = df.drop(columns=["RowNumber", "CustomerId", "Surname"], errors="ignore")
    
    # One-hot encode categorical variables
    df = pd.get_dummies(df, columns=["Geography", "Gender"], drop_first=True)
    
    return df

In [10]:
# Load the dataset
df = pd.read_csv("datasets/Churn_Modelling_train_test.csv")

processed_df = transform(df)
processed_df.head()

,CreditScore,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited,Geography_Germany,Geography_Spain,Gender_Male
0,710,37.0,5,0.00,2,1.0,0.0,115403.31,0,False,False,False
1,569,38.0,8,0.00,2,0.0,0.0,79618.79,0,False,True,True
2,710,54.0,6,171137.62,1,1.0,1.0,167023.95,1,False,False,True
3,673,43.0,8,127132.96,1,0.0,1.0,6009.27,1,False,True,True
4,610,69.0,5,86038.21,3,0.0,0.0,192743.06,1,True,False,False


In [11]:
# balance the dataset (if you think it is necessary)
df["Exited"].value_counts(normalize=True) * 100

Exited
0    79.557827
1    20.442173
Name: proportion, dtype: float64

The dataset should be treated as imbalanced because the target variable `Exited` contains more non-churned customers than churned customers. This means that accuracy alone may be misleading, as a model could achieve high accuracy by mostly predicting the majority class. To address this, I used `class_weight="balanced"` in the Decision Tree model. This allows the model to give more weight to the minority class and improves its ability to identify customers who actually churned.

In [12]:
from sklearn.tree import DecisionTreeClassifier

model = DecisionTreeClassifier(
    max_depth=5,
    random_state=42,
    class_weight="balanced"
)

In [13]:
from sklearn.model_selection import train_test_split

# Split the processed dataset into features and target
X = processed_df.drop(columns=["Exited"])
y = processed_df["Exited"]

# Create train and test samples
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train distribution:")
print(y_train.value_counts(normalize=True))
print("y_test distribution:")
print(y_test.value_counts(normalize=True))

X_train shape: (7200, 11)
X_test shape: (1801, 11)
y_train distribution:
Exited
0    0.795556
1    0.204444
Name: proportion, dtype: float64
y_test distribution:
Exited
0    0.795669
1    0.204331
Name: proportion, dtype: float64


In [14]:
# Train a decision tree model using the `train sample`

model.fit(X_train, y_train)

DecisionTreeClassifier(class_weight='balanced', max_depth=5, random_state=42)

In [15]:
from sklearn.metrics import accuracy_score, f1_score, classification_report, confusion_matrix

# Predict on the test sample
y_pred = model.predict(X_test)

# Calculate metrics
accuracy = accuracy_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

print("Accuracy:", accuracy)
print("F1 Score:", f1)

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred))

print("\nClassification Report:")
print(classification_report(y_test, y_pred))

Accuracy: 0.691282620766241
F1 Score: 0.513986013986014

Confusion Matrix:
[[951 482]
 [ 74 294]]

Classification Report:
              precision    recall  f1-score   support

           0       0.93      0.66      0.77      1433
           1       0.38      0.80      0.51       368

    accuracy                           0.69      1801
   macro avg       0.65      0.73      0.64      1801
weighted avg       0.82      0.69      0.72      1801



## Model Evaluation

The Decision Tree model achieved an accuracy of approximately 69.1% and an F1 score of approximately 0.51. Since the dataset is imbalanced, accuracy alone is not the best measure of performance.

The confusion matrix shows that the model correctly predicted 951 non-churned customers and 294 churned customers. It incorrectly predicted 482 customers as churned when they did not churn, and incorrectly predicted 74 customers as non-churned when they actually churned.

For the churn class (`Exited = 1`), the model achieved a recall of 0.80. This means the model identified 80% of the customers who actually churned, which is useful in a churn prediction problem. However, the precision for churn was only 0.38, meaning many customers predicted as churned were false positives.

Overall, the model is useful for identifying many potential churners, but it produces a high number of false alarms. This may be acceptable if the business prefers to catch more churned customers, even if some non-churned customers are incorrectly flagged.

### MLFlow Integration

In this section the students are required to track the model and the parameters into MLFlow.

In [16]:
import mlflow
from mlflow.models import infer_signature

mlflow.set_tracking_uri("http://127.0.0.1:8080")
mlflow.set_experiment("Practice Experiment - Aaron")

2026/05/14 20:16:47 INFO mlflow.tracking.fluent: Experiment with name 'Practice Experiment - Aaron' does not exist. Creating a new experiment.


<Experiment: artifact_location='mlflow-artifacts:/126766513514206643', creation_time=1778782607240, experiment_id='126766513514206643', last_update_time=1778782607240, lifecycle_stage='active', name='Practice Experiment - Aaron', tags={}>

Start a local Tracking server on port 8080 - it is necessary to import mlflow and open a new terminal

In [17]:
# Set our tracking server uri for logging on port 8080

mlflow.set_tracking_uri("http://127.0.0.1:8080")

Visit the MLFlow UI on `http://127.0.0.1:8080`

In [19]:
# Create a new MLflow Experiment called `Practice Experiment - Aaron`

mlflow.set_experiment("Practice Experiment - Aaron Glanz")

2026/05/14 20:20:05 INFO mlflow.tracking.fluent: Experiment with name 'Practice Experiment - Aaron Glanz' does not exist. Creating a new experiment.


<Experiment: artifact_location='mlflow-artifacts:/796188224937888532', creation_time=1778782805265, experiment_id='796188224937888532', last_update_time=1778782805265, lifecycle_stage='active', name='Practice Experiment - Aaron Glanz', tags={}>

In [21]:
from mlflow.models import infer_signature

params = {
    "max_depth": 5,
    "random_state": 42,
    "class_weight": "balanced"
}

with mlflow.start_run():
    # Log model hyperparameters
    mlflow.log_params(params)

    # Log model metrics
    mlflow.log_metric("accuracy", accuracy)
    mlflow.log_metric("f1_score", f1)

    # Add useful tags
    mlflow.set_tag("model_type", "Decision Tree")
    mlflow.set_tag("task", "Customer churn prediction")
    mlflow.set_tag("dataset", "Churn_Modelling_train_test.csv")

    # Infer model signature
    signature = infer_signature(X_train, model.predict(X_train))

    # Log the trained model
    mlflow.sklearn.log_model(
        sk_model=model,
        artifact_path="decision_tree_churn_model",
        signature=signature,
        input_example=X_train.head()
    )

/Users/aaronglanz/Documents/GitHub/mlops-and-system-design/.venv/lib/python3.13/site-packages/mlflow/types/utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details.
  warnings.warn(


🏃 View run amazing-mule-151 at: http://127.0.0.1:8080/#/experiments/796188224937888532/runs/b3769083e4e34a27a84c39df17ff5eb2
🧪 View experiment at: http://127.0.0.1:8080/#/experiments/796188224937888532


### Keep practicing

Experiment 2: Decision Tree with different hyperparameters

In [ ]:



model_2 = DecisionTreeClassifier(
    max_depth=8,
    min_samples_split=20,
    min_samples_leaf=10,
    random_state=42,
    class_weight="balanced"
)

model_2.fit(X_train, y_train)

y_pred_2 = model_2.predict(X_test)

accuracy_2 = accuracy_score(y_test, y_pred_2)
f1_2 = f1_score(y_test, y_pred_2)

print("Accuracy:", accuracy_2)
print("F1 Score:", f1_2)

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_2))

print("\nClassification Report:")
print(classification_report(y_test, y_pred_2))

Accuracy: 0.7357023875624653
F1 Score: 0.5296442687747036

Confusion Matrix:
[[1057  376]
 [ 100  268]]

Classification Report:
              precision    recall  f1-score   support

           0       0.91      0.74      0.82      1433
           1       0.42      0.73      0.53       368

    accuracy                           0.74      1801
   macro avg       0.66      0.73      0.67      1801
weighted avg       0.81      0.74      0.76      1801



In [23]:
params_2 = {
    "max_depth": 8,
    "min_samples_split": 20,
    "min_samples_leaf": 10,
    "random_state": 42,
    "class_weight": "balanced"
}

with mlflow.start_run(run_name="decision_tree_experiment_2"):
    # Log hyperparameters
    mlflow.log_params(params_2)

    # Log metrics
    mlflow.log_metric("accuracy", accuracy_2)
    mlflow.log_metric("f1_score", f1_2)

    # Add tags
    mlflow.set_tag("model_type", "Decision Tree")
    mlflow.set_tag("task", "Customer churn prediction")
    mlflow.set_tag("experiment_note", "Deeper tree with minimum sample constraints")

    # Infer model signature
    signature_2 = infer_signature(X_train, model_2.predict(X_train))

    # Log model
    mlflow.sklearn.log_model(
        sk_model=model_2,
        artifact_path="decision_tree_churn_model_experiment_2",
        signature=signature_2,
        input_example=X_train.head()
    )

/Users/aaronglanz/Documents/GitHub/mlops-and-system-design/.venv/lib/python3.13/site-packages/mlflow/types/utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details.
  warnings.warn(


🏃 View run decision_tree_experiment_2 at: http://127.0.0.1:8080/#/experiments/796188224937888532/runs/7f6a7717d43041f2952442e75f059b14
🧪 View experiment at: http://127.0.0.1:8080/#/experiments/796188224937888532


Experiment 3: Decision tree without balancing


In [24]:
model_3 = DecisionTreeClassifier(
    max_depth=5,
    random_state=42
)

model_3.fit(X_train, y_train)

y_pred_3 = model_3.predict(X_test)

accuracy_3 = accuracy_score(y_test, y_pred_3)
f1_3 = f1_score(y_test, y_pred_3)

print("Accuracy:", accuracy_3)
print("F1 Score:", f1_3)
print(confusion_matrix(y_test, y_pred_3))
print(classification_report(y_test, y_pred_3))

Accuracy: 0.8561910049972238
F1 Score: 0.5350089766606823
[[1393   40]
 [ 219  149]]
              precision    recall  f1-score   support

           0       0.86      0.97      0.91      1433
           1       0.79      0.40      0.54       368

    accuracy                           0.86      1801
   macro avg       0.83      0.69      0.72      1801
weighted avg       0.85      0.86      0.84      1801



In [25]:
params_3 = {
    "max_depth": 5,
    "random_state": 42,
    "class_weight": None
}

with mlflow.start_run(run_name="decision_tree_no_class_weight"):
    mlflow.log_params(params_3)
    mlflow.log_metric("accuracy", accuracy_3)
    mlflow.log_metric("f1_score", f1_3)

    mlflow.set_tag("model_type", "Decision Tree")
    mlflow.set_tag("experiment_note", "Decision tree without class balancing")

    signature_3 = infer_signature(X_train, model_3.predict(X_train))

    mlflow.sklearn.log_model(
        sk_model=model_3,
        artifact_path="decision_tree_no_class_weight",
        signature=signature_3,
        input_example=X_train.head()
    )

/Users/aaronglanz/Documents/GitHub/mlops-and-system-design/.venv/lib/python3.13/site-packages/mlflow/types/utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details.
  warnings.warn(


🏃 View run decision_tree_no_class_weight at: http://127.0.0.1:8080/#/experiments/796188224937888532/runs/28d17f90d06d4584964ac8e828826e38
🧪 View experiment at: http://127.0.0.1:8080/#/experiments/796188224937888532


Experiment 4: Shallower Decision tree 

In [26]:
model_4 = DecisionTreeClassifier(
    max_depth=3,
    random_state=42,
    class_weight="balanced"
)

model_4.fit(X_train, y_train)

y_pred_4 = model_4.predict(X_test)

accuracy_4 = accuracy_score(y_test, y_pred_4)
f1_4 = f1_score(y_test, y_pred_4)

print("Accuracy:", accuracy_4)
print("F1 Score:", f1_4)
print(confusion_matrix(y_test, y_pred_4))
print(classification_report(y_test, y_pred_4))

Accuracy: 0.6785119378123264
F1 Score: 0.48533333333333334
[[949 484]
 [ 95 273]]
              precision    recall  f1-score   support

           0       0.91      0.66      0.77      1433
           1       0.36      0.74      0.49       368

    accuracy                           0.68      1801
   macro avg       0.63      0.70      0.63      1801
weighted avg       0.80      0.68      0.71      1801



In [27]:
params_4 = {
    "max_depth": 3,
    "random_state": 42,
    "class_weight": "balanced"
}

with mlflow.start_run(run_name="decision_tree_shallow"):
    mlflow.log_params(params_4)
    mlflow.log_metric("accuracy", accuracy_4)
    mlflow.log_metric("f1_score", f1_4)

    mlflow.set_tag("model_type", "Decision Tree")
    mlflow.set_tag("experiment_note", "Shallower tree to reduce overfitting")

    signature_4 = infer_signature(X_train, model_4.predict(X_train))

    mlflow.sklearn.log_model(
        sk_model=model_4,
        artifact_path="decision_tree_shallow",
        signature=signature_4,
        input_example=X_train.head()
    )

/Users/aaronglanz/Documents/GitHub/mlops-and-system-design/.venv/lib/python3.13/site-packages/mlflow/types/utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details.
  warnings.warn(


🏃 View run decision_tree_shallow at: http://127.0.0.1:8080/#/experiments/796188224937888532/runs/c368214870fb4edfa4e06279a877a0f4
🧪 View experiment at: http://127.0.0.1:8080/#/experiments/796188224937888532


Experiment 5: Random Forest

In [28]:
from sklearn.ensemble import RandomForestClassifier

model_5 = RandomForestClassifier(
    n_estimators=100,
    max_depth=8,
    random_state=42,
    class_weight="balanced"
)

model_5.fit(X_train, y_train)

y_pred_5 = model_5.predict(X_test)

accuracy_5 = accuracy_score(y_test, y_pred_5)
f1_5 = f1_score(y_test, y_pred_5)

print("Accuracy:", accuracy_5)
print("F1 Score:", f1_5)
print(confusion_matrix(y_test, y_pred_5))
print(classification_report(y_test, y_pred_5))

Accuracy: 0.8195446973903387
F1 Score: 0.6060606060606061
[[1226  207]
 [ 118  250]]
              precision    recall  f1-score   support

           0       0.91      0.86      0.88      1433
           1       0.55      0.68      0.61       368

    accuracy                           0.82      1801
   macro avg       0.73      0.77      0.74      1801
weighted avg       0.84      0.82      0.83      1801



In [29]:
params_5 = {
    "model": "RandomForestClassifier",
    "n_estimators": 100,
    "max_depth": 8,
    "random_state": 42,
    "class_weight": "balanced"
}

with mlflow.start_run(run_name="random_forest_balanced"):
    mlflow.log_params(params_5)
    mlflow.log_metric("accuracy", accuracy_5)
    mlflow.log_metric("f1_score", f1_5)

    mlflow.set_tag("model_type", "Random Forest")
    mlflow.set_tag("task", "Customer churn prediction")
    mlflow.set_tag("experiment_note", "Random forest model with balanced class weights")

    signature_5 = infer_signature(X_train, model_5.predict(X_train))

    mlflow.sklearn.log_model(
        sk_model=model_5,
        artifact_path="random_forest_churn_model",
        signature=signature_5,
        input_example=X_train.head()
    )

/Users/aaronglanz/Documents/GitHub/mlops-and-system-design/.venv/lib/python3.13/site-packages/mlflow/types/utils.py:452: UserWarning: Hint: Inferred schema contains integer column(s). Integer columns in Python cannot represent missing values. If your input data contains missing values at inference time, it will be encoded as floats and will cause a schema enforcement error. The best way to avoid this problem is to infer the model schema based on a realistic data sample (training dataset) that includes missing values. Alternatively, you can declare integer columns as doubles (float64) whenever these columns may have missing values. See `Handling Integers With Missing Values <https://www.mlflow.org/docs/latest/models.html#handling-integers-with-missing-values>`_ for more details.
  warnings.warn(


🏃 View run random_forest_balanced at: http://127.0.0.1:8080/#/experiments/796188224937888532/runs/fff5571b414d460eb272e7073f4817b9
🧪 View experiment at: http://127.0.0.1:8080/#/experiments/796188224937888532


Random Forest is the best performing mdoel on the test data so far